netsim/modules/dhcp.py

```diff
@@ -118,38 +118,93 @@ def check_protocol_support(node: Box, topology: Box) -> bool:
+      pools[pid].vrf = vdata.vrf
+
+  # Phase 2: Iterate over node interfaces, find SVIs, and build excluded IP list
+  #
+  for node,ndata in topology.get('nodes',{}).items():
+    for intf in ndata.get('interfaces',[]):
+      vname = intf.get('vlan.name')
+      if not vname:
+        continue
+      pid = strings.make_id(f'vlan_{vname}')
+
+      if pid not in pools:
+        continue
+
+      # Iterate over all address families, finding true interface addresses,
+      # converting them from CIDR format to IPv4/IPv6 address format, and
+      # appending them to the pool.excluded.af list
+      #
+      for af in log.AF_LIST:
+        if af in intf and isinstance(intf[af],str):
+          data.append_to_list(pools[pid].excluded,af,str(netaddr.IPNetwork(intf[af]).ip))
 
   for link in topology.get('links',[]):                     # Iterate over lab topology links
     if not link.get('dhcp.subnet'):                         # dhcp.subnet is set if there's at least one DHCP client on the link
       continue
 
-    subnet = data.get_empty_box()                           # Create a DHCP pool data structure
-    if 'vrf' in link:                                       # Copy link VRF information into the pool
-      subnet.vrf = link.vrf                                 # ... to support VRF-aware DHCP servers
-
-    for af in ('ipv4','ipv6'):                              # Iterate over link address families
+    vname = link.get('vlan.access',None)
+    lname = link.get('name','') or link.get('_linkname','')
+    if vname:
+      pid = strings.make_id(f'vlan_{vname}')
+    else:
+      pid = strings.make_id(lname)
+      pools[pid].name = lname
+      if 'vrf' in link:                                     # Copy VRF information from non-VLAN links
+        pools[pid].vrf = link.vrf                           # ... to support VRF-aware DHCP servers
+
+    pools[pid].active = True
+    for af in log.AF_LIST:                                  # Iterate over link address families
       if af not in link.dhcp.subnet or af not in link.prefix:
         continue                                            # No AF prefix or no clients within this AF
 
-      subnet[af] = link.prefix[af]                          # Copy link prefix into the pool
-      subnet.name = link.get('name','') or link.get('_linkname','')
-      subnet.clean_name = strings.make_id(subnet.name)      # Get pool name from link and clean it up
+      if af in pools[pid]:                                  # Pool already has a prefix
+        if pools[pid][af] != link.prefix[af]:               # Check for mismatch between link and VLAN prefix
+          log.error(
+            f'Mismatch in DHCP pool {pid} prefix {pools[pid][af]}, link {lname} claims the prefix should be {link.prefix[af]}',
+            category=log.IncorrectValue,
+            module='dhcp')
+      else:
+        pools[pid][af] = link.prefix[af]                    # New pool, add prefix
```

Comment:
Could throw an error if link.prefix[af] is False

Reply:
Another good one. It would definitely mess up the DHCP configuration template (or confuse the DHCP daemon). I added a test earlier on whether the prefix is a string (because "True" doesn't help either).